framework/src/kernels/KernelBase.C

```diff
@@ -34,6 +34,7 @@ InputParameters validParams<KernelBase>()
   params += validParams<MaterialPropertyInterface>();
 
   params.addRequiredParam<NonlinearVariableName>("variable", "The name of the variable that this Kernel operates on");
+  params.addParam<bool>("calculate_nonlocal_variable_jacobian", false, "Whether or not this object requires calculating jacobian contribution corresponding to non-local variables");
```

Comment:
Well, since I am making changes in the framework, I kept the default false so that user can turn it on when needed. "detect" sounds better though. I will see if I can assign a flag in the UserObject to make it detect automatically.